### PR TITLE
fix(resourcehandler): let Kustomize own referenced Helm charts

### LINKED ISSUE #2853: [bug]: Kustomize helmCharts are also rendered as standalone Helm charts
# Description

When a scanned Kustomize configuration uses `helmCharts`, the referenced local chart is rendered twice through two independent providers.

`getResourcesFromPath` first runs the generic Helm loader recursively over the input tree. It later builds the Kustomize directory, whose Helm chart inflator renders the same chart again and then applies Kustomize transformations.

With a chart that emits `ConfigMap/app` and a Kustomization containing `namePrefix: prod-`, the returned workload set contains both `ConfigMap/app` from the generic Helm render and `ConfigMap/prod-app` from Kustomize. The standalone object is not part of the Kustomize build.

This report concerns ownership between two successful renderers. It is not about parsing raw Go-template files as YAML.

# Environment

OS: `Ubuntu 24.04.3 LTS`  
Version: `master` at [`10e16a2a4e3011d6b25ca06bbe3caee7a7ee71ac`](https://github.com/kubescape/kubescape/commit/10e16a2a4e3011d6b25ca06bbe3caee7a7ee71ac)  
Go: `go1.26.5 linux/amd64`  
Helm: deterministic test double reporting `v3.14.0+gtest` (shown below)

# Steps To Reproduce

Add this regression test to `core/pkg/resourcehandler/filesloader_test.go`:

```go
func TestGetResourcesFromPath_KustomizeOwnsReferencedHelmChart(t *testing.T) {
    root := t.TempDir()
    chartDir := filepath.Join(root, "vendor", "charts", "app")
    templateDir := filepath.Join(chartDir, "templates")
    require.NoError(t, os.MkdirAll(templateDir, 0o750))
    require.NoError(t, os.WriteFile(filepath.Join(root, "kustomization.yaml"), []byte(`apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namePrefix: prod-
helmGlobals:
  chartHome: vendor/charts
helmCharts:
  - name: app
    releaseName: app
`), 0o600))
    require.NoError(t, os.WriteFile(
        filepath.Join(chartDir, "Chart.yaml"),
        []byte("apiVersion: v2\nname: app\nversion: 0.1.0\n"), 0o600,
    ))
    require.NoError(t, os.WriteFile(
        filepath.Join(chartDir, "values.yaml"), []byte("{}\n"), 0o600,
    ))
    require.NoError(t, os.WriteFile(
        filepath.Join(templateDir, "configmap.yaml"),
        []byte(`apiVersion: v1
kind: ConfigMap
metadata:
  name: app
`), 0o600,
    ))

    _, workloads, err := getResourcesFromPath(
        context.Background(), root, cautils.HelmValueOptions{},
    )
    require.NoError(t, err)

    counts := map[string]int{}
    for _, workload := range workloads {
        counts[workload.GetKind()+"/"+workload.GetName()]++
    }
    assert.Equal(t, 1, counts["ConfigMap/prod-app"])
    assert.Zero(t, counts["ConfigMap/app"])
}
```

Use this deterministic `helm` test double. It implements the two commands the Kustomize Helm inflator uses in this fixture and returns the chart's valid rendered manifest:

```sh
mkdir -p /tmp/kubescape-fake-helm
cat >/tmp/kubescape-fake-helm/helm <<'EOF'
#!/bin/sh
set -eu
case "${1:-}" in
  version)
    printf '%s\n' 'v3.14.0+gtest'
    ;;
  template)
    shift
    shift # release name
    chart_path=$1
    cat "$chart_path/templates/configmap.yaml"
    ;;
  *)
    exit 2
    ;;
esac
EOF
chmod 0755 /tmp/kubescape-fake-helm/helm
```

Run:

```bash
PATH="/tmp/kubescape-fake-helm:$PATH" go test ./core/pkg/resourcehandler \
  -run '^TestGetResourcesFromPath_KustomizeOwnsReferencedHelmChart$' \
  -count=3
```

On current `master`, all three runs fail the standalone-object assertion:

```text
Should be zero, but was 1
TestGetResourcesFromPath_KustomizeOwnsReferencedHelmChart
```

The preceding `prod-app` assertion passes. Inspecting `counts` shows one `ConfigMap/prod-app` and one `ConfigMap/app` on every run.

# Expected behavior

A chart listed in the selected Kustomization's `helmCharts` belongs to the Kustomize build. Only the Kustomize-transformed `ConfigMap/prod-app` should enter the scan.

An unrelated Helm chart under the same scan tree that is not referenced by that Kustomization should continue to be rendered by the generic Helm loader.

# Actual Behavior

The generic recursive Helm loader renders the local chart as a standalone chart. The Kustomize renderer then renders it again and applies `namePrefix`, leaving both the standalone and transformed objects in the returned workload set.

# Root cause

`getResourcesFromPath` invokes the generic Helm loader before Kustomize:

```go
helmSourceToWorkloads, helmSourceToChart, renderedCharts, err :=
    cautils.LoadResourcesFromHelmCharts(ctx, path, helmValueOpts)
```

and later independently invokes:

```go
kustomizeSourceToWorkloads, kustomizeDirectoryName, err :=
    cautils.LoadResourcesFromKustomizeDirectory(ctx, path)
```

`listHelmChartDirs` walks recursively and has no knowledge of `helmCharts` ownership, so it claims `<chartHome>/<chart.name>` before Kustomize builds it.

The final identity deduplication is not sufficient:

- when Kustomize changes the name or namespace, the two outputs have different identities;
- without an identity transformation, Helm and Kustomize are both rank-two rendered providers, and same-rank duplicates are intentionally retained.

# Impact

The scan can include a standalone Helm object that the selected Kustomize configuration does not build. It may use different release identity, values, namespace, or omit Kustomize transformations and patches. That extra object can be evaluated and reported alongside the intended Kustomize output.

The reproducer establishes the input-composition error; it does not claim a particular policy verdict.

# Suggested fix

Before generic Helm rendering:

- parse the selected Kustomization using Kustomize's own `types.Kustomization`;
- follow only its explicit local `resources` and `components`, because those configurations participate in the same selected build and can define their own `helmCharts`;
- resolve `helmGlobals.chartHome` (including the default `charts`) and each explicitly referenced `helmCharts[].name`;
- mirror Kustomize's `<name>-<version>/<name>` storage path for versioned repository charts;
- exclude those chart directories and their subcharts from the generic Helm render;
- still exclude their templates from the plain-YAML pass because Kustomize owns them;
- leave unrelated sibling charts to the generic Helm loader.

Add:

- a provider-level test proving that a referenced chart and its subcharts are excluded while an unreferenced sibling chart remains;
- a selected-root/base/component traversal test, without scanning for unrelated Kustomizations elsewhere in the repository;
- an integration test proving that only the Kustomize-transformed object is returned with Kustomize source attribution.

# Additional context

PR #2509 fixed a different overlap: successfully rendered Helm templates were also parsed as plain YAML, causing warnings and Helm-versus-plain duplication. It excludes raw `templates/` files after a generic Helm render. It does not prevent a chart referenced by Kustomize from being rendered independently by both Helm and Kustomize.

I searched open issues and open pull requests for the exact title and for Kustomize/`helmCharts` duplicate or standalone rendering. I could not find an existing report or in-flight change for this failure mode.

### PR DESCRIPTION
**Pull request by ye11oc4t**
Resolves #2853

## Summary

- resolve the physical chart directories explicitly referenced by the selected Kustomization and its local resource/component graph;
- leave those charts and their nested subcharts out of the generic recursive Helm renderer;
- keep their raw templates out of the plain-manifest pass because the Kustomize render owns them;
- preserve generic Helm rendering for unrelated sibling charts.

## Root cause

The local path loader independently ran a recursive Helm render and a Kustomize build. A local chart referenced through `helmCharts` was therefore rendered once as a standalone Helm release and once inside Kustomize.

The existing final deduplication cannot repair this reliably. Kustomize transformations can change the resource identity, and same-rank rendered-provider duplicates are retained intentionally.

## Implementation

`KustomizeHelmChartDirectories` parses the selected Kustomization with `types.Kustomization.Unmarshal`, applies `FixKustomization` for deprecated fields, and follows only explicit, existing local `resources` and `components`. A cycle-safe visited set keeps this traversal bounded to the selected build graph; it does not search the repository for other Kustomizations.

For each configuration it resolves the default or configured chart home and the physical roots explicitly named by `helmCharts`. Existing chart roots are expanded through their nested `Chart.yaml` directories so both generic Helm discovery and the raw-template exclusion account for subcharts.

For charts that specify both `repo` and `version`, it mirrors Kustomize's own downloaded-chart layout, `<chartHome>/<name>-<version>/<name>`, so a previously pulled chart is assigned to the same renderer that consumes it.

`LoadResourcesFromHelmChartsExcludingDirectories` filters those roots before constructing or rendering generic Helm charts. Directory containment compares absolute canonical paths for existing directories, while retaining an absolute lexical fallback for remote charts that have not been pulled yet. Template-file exclusion accepts either lexical or canonical containment so a symlinked template whose target is outside the chart is still owned by Helm/Kustomize.

Nested chart discovery remains best effort: successfully discovered chart roots are retained, while per-path `stat` and walk failures are logged with the same warning contract as generic Helm discovery instead of aborting the whole scan.

The plain-file loader receives both successfully generic-rendered chart roots and Kustomize-owned chart roots.

## Tests

Provider-level coverage creates a selected root that references a local base:

```text
root/
├── kustomization.yaml       # resources: [base]
├── base/
│   ├── kustomization.yaml   # helmCharts: [app]
│   └── charts/app/
│       └── charts/dependency/
└── charts/standalone/       # not referenced
```

It asserts that only `standalone` reaches the generic Helm result and that both the referenced parent and nested chart templates are excluded from the raw-manifest pass.

Parser-level cases cover a custom and absolute chart home, an explicit local resource/component graph, repository-only and version-only path semantics, two release entries that reference the same physical chart, and Kustomize's versioned repository-chart storage path.

The end-to-end file-loader regression installs a strict test-local Helm executable, inflates one physical chart twice (`releaseName: first` and `second`) under `namePrefix: prod-`, and asserts:

- `ConfigMap/prod-first` and `ConfigMap/prod-second` each occur exactly once;
- there is no third standalone ConfigMap;
- both surviving objects have Kustomize source attribution.

Additional regressions cover a symlinked scan root, a symlinked template leaf, cwd-relative scan input, and partial nested-chart discovery errors.

## Verification

A clean `10e16a2a` baseline with the issue reproducer failed deterministically on all three runs because `ConfigMap/app` remained. The same deterministic Helm contract was used for the fixed tests.

```bash
go test ./core/cautils \
  -run '^(TestKustomizeHelmChartDirectories.*|TestLoadResourcesFromHelmChartsExcludingKustomizeOwnedDirectories)$' \
  -count=3
go test ./core/pkg/resourcehandler \
  -run '^TestGetResourcesFromPath_KustomizeOwnsReferencedHelmChart$' \
  -count=3
go test ./core/cautils ./core/pkg/resourcehandler
go vet ./core/cautils ./core/pkg/resourcehandler
```

All four fixed commands passed. Removing the ownership lookup made the integration test fail with three ConfigMaps instead of two and attributed the extra object to the Helm provider; disabling canonical path comparison made the symlink regression fail. Restoring both changes returned the focused suite to green.

Both a full affected-package race run and a targeted ownership race run were capped after 300 seconds of compilation in this constrained environment. Neither produced a race diagnostic before timeout; remote race CI remains required. The integration test no longer has an availability-based skip path: its test-local Helm double accepts only the exact commands used by this fixture and fails on command drift.

## Scope

This assigns ownership only for charts explicitly listed by the Kustomization selected directly by the scan input and configurations reached through that build's local `resources`/`components`. It does not discover unrelated nested Kustomizations, suppress the whole chart home, reinterpret unrelated charts, or change Helm-versus-plain behavior from #2509.

When composed with recursive nested-Kustomize discovery, discovery must run first and the union of referenced chart roots from every configuration being rendered must be supplied to the Helm exclusion. This PR keeps that boundary explicit rather than independently choosing which nested Kustomizations represent target environments.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling of Helm charts referenced by Kustomize.
  * Prevented Kustomize-managed charts from being rendered or loaded again as standalone Helm releases.
  * Preserved multiple Kustomize inflations of the same Helm chart with correct source attribution.
  * Added support for excluding specified chart directories during Helm resource loading.
  * Improved chart discovery across symlinked, nested, and versioned paths.

### CODE REVIEW

File: core/cautils/fileutils.go [hunk 1]
```diff
@@ -51,10 +51,31 @@ type Chart struct {
...
 	}
+	if len(excludedChartDirectories) > 0 {
+		remaining := make([]string, 0, len(helmDirectories))
+		for _, directory := range helmDirectories {
+			if !isUnderAnyDir(directory, excludedChartDirectories) {
+				remaining = append(remaining, directory)
+			}
+		}
+		helmDirectories = remaining
+	}
```
[the remark this answers is not in the document]
Comment (ye11oc4t) [reply core/cautils/fileutils.go:79]: Thanks for flagging this. I traced `listHelmChartDirs` → `listDirs` → `listFilesOrDirectories`: a non-absolute scan pattern is converted with `filepath.Abs` before the walk, so this call site already returns absolute discovered chart paths. I added a regression that runs the ownership exclusion with both an absolute root and a cwd-relative root; both pass. The containment code also normalizes/canonicalizes both discovered and excluded directories before comparison, with the excluded side precomputed once. This covers physical path aliases without relying on the relative-path assumption.

File: core/cautils/kustomizedirectory.go [hunk 2]
```diff
@@ -70,6 +72,159 @@ func getKustomizeDirectoryName(path string) string {
...
+	seenChartDirectories map[string]struct{},
+	chartDirectories *[]string,
+) error {
+	absKustomizationPath, err := filepath.Abs(kustomizationPath)
+	if err != nil {
+		return fmt.Errorf("failed to resolve Kustomization path %q: %w", kustomizationPath, err)
+	}
+	if resolved, err := filepath.EvalSymlinks(absKustomizationPath); err == nil {
+		absKustomizationPath = resolved
+	}
```
[the remark this answers is not in the document]
Comment (ye11oc4t) [reply core/cautils/kustomizedirectory.go]: Agreed; this was a real aliasing gap, fixed in `6fa0fcd`. Existing paths now use `filepath.Abs` plus `filepath.EvalSymlinks`; missing remote-chart paths retain their absolute lexical form, while other resolution errors follow the caller’s existing error/log contract. Kustomization paths, owned chart roots, nested charts, generic discoveries, and template exclusions now compare consistent forms. I added an intermediate-parent-symlink regression. A final hostile pass also caught the inverse edge—a template-file symlink targeting outside the chart—so template ownership now accepts either lexical or canonical containment, with a focused regression.

File: core/cautils/kustomizedirectory.go [hunk 3]
```diff
@@ -70,6 +72,159 @@ func getKustomizeDirectoryName(path string) string {
...
+		return fmt.Errorf("failed to inspect Kustomize-owned Helm chart %q: %w", directory, err)
+	}
+	if !info.IsDir() {
+		return nil
+	}
+
+	nestedDirectories, discoveryErrs := listHelmChartDirs(directory)
+	if len(discoveryErrs) > 0 {
+		return fmt.Errorf("failed to discover Kustomize-owned Helm chart dependencies below %q: %w", directory, errors.Join(discoveryErrs...))
+	}
```
[the remark this answers is not in the document]
Comment (ye11oc4t) [reply core/cautils/kustomizedirectory.go]: Fixed in `6fa0fcd` while preserving the generic Helm loader’s best-effort contract. Read/parse errors for the selected Kustomization remain hard failures because the owning Kustomize render cannot succeed, but nested owned-chart `stat`/walk errors now emit the same `Skipping path while discovering Helm charts` warning and continue. Successfully discovered nested directories are retained even when the walk also returns errors. The regression injects one nested directory plus a synthetic discovery error and asserts that both the root and partial result remain owned.

File: core/pkg/resourcehandler/filesloader_test.go [hunk 4]
```diff
@@ -251,6 +251,65 @@ func TestGetResourcesFromPath_KustomizeTransformersDoNotDuplicate(t *testing.T)
...
+	require.NoError(t, os.WriteFile(filepath.Join(templateDir, "configmap.yaml"), []byte(`apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ .Release.Name }}
+`), 0o600))
+
+	sources, workloads, err := getResourcesFromPath(context.Background(), root, cautils.HelmValueOptions{})
+	if err != nil && kustomizeHelmUnavailable(err) {
+		t.Skipf("Kustomize Helm integration is unavailable: %v", err)
+	}
```
[the remark this answers is not in the document]
Comment (ye11oc4t) [reply core/pkg/resourcehandler/filesloader_test.go]: Thanks. I did not add a workflow-level Helm installation: run 31269031748 already shows this exact regression executing and passing, rather than skipping. I hardened it further in `6fa0fcd`: the test installs a strict local `helm`, sets `PATH` to that temporary directory, accepts only Kustomize’s exact `version -c --short` and `template` contracts, verifies the chart files, emits deterministic manifests, and rejects every other invocation. The availability-based skip helper is gone, so CI cannot silently skip this path.

**Comment by ye11oc4t** [comment]
Follow-up on the outside-diff CRD composition concern: the direct input paths changed by this PR do not reach the recursive plain-file pass, so the reported raw-CRD duplication is not reachable in #2855 itself. It does become reachable when repository-root nested Kustomize discovery is added. Draft PR #2860 handles that composition explicitly: it excludes raw chart CRDs only when the successful `helmCharts` render uses `includeCRDs: true`, retains the raw CRD when `false`, and asserts the rendered/raw partition before final identity dedup. This keeps the direct ownership fix scoped while covering the first reachable broad-scan path.

**Review by matthyx — APPROVED** [review]
LGTM! The implementation cleanly resolves the duplication issue when Kustomize references Helm charts. Unit tests are thorough, edge cases (chart home layouts, graph traversal, partial walks) are covered, and CI is green.

**Comment by matthyx** [comment]
@ye11oc4t is it ready?

**Comment by ye11oc4t** [comment]
@matthyx   Yes, everything is ready now. Thanks for the review!
